File: server/utils/TokenManager.js
/**
 * Gestionnaire de tokens avec expiration, rotation et révocation
 */

import crypto from 'crypto'

class TokenManager {
  constructor(defaultTtl = 24 * 60 * 60 * 1000) {
    // defaultTtl: 24 heures par défaut
    this.tokens = new Map() // token -> { issuedAt, expiresAt, clientId, metadata }
    this.revokedTokens = new Set() // Tokens révoqués
    this.defaultTtl = defaultTtl
    this.cleanupInterval = setInterval(() => this.cleanup(), 60 * 60 * 1000) // Nettoyage toutes les heures
  }

  /**
   * Génère un nouveau token
   */
  issueToken(clientId = null, metadata = {}, customTtl = null) {
    const token = this.generateToken()
    const now = Date.now()
    const ttl = customTtl || this.defaultTtl
    const expiresAt = now + ttl

    this.tokens.set(token, {
      issuedAt: now,
      expiresAt,
      clientId,
      metadata,
      lastUsed: now,
    })

    return {
      token,
      expiresAt,
      expiresIn: ttl,
    }
  }

  /**
   * Vérifie si un token est valide
   */
  isValid(token) {
    if (!token) return false

    // Vérifier si le token est révoqué
    if (this.revokedTokens.has(token)) {
      return false
    }

    const tokenData = this.tokens.get(token)
    if (!tokenData) {
      return false
    }

    // Vérifier l'expiration
    if (Date.now() > tokenData.expiresAt) {
      this.tokens.delete(token)
      return false
    }

    // Mettre à jour la dernière utilisation
    tokenData.lastUsed = Date.now()

    return true
  }

  /**
   * Révoke un token
   */
  revokeToken(token) {
    if (this.tokens.has(token)) {
      this.tokens.delete(token)
    }
    this.revokedTokens.add(token)
  }

  /**
   * Révoke tous les tokens d'un client
   */
  revokeClientTokens(clientId) {
    for (const [token, data] of this.tokens.entries()) {
      if (data.clientId === clientId) {
        this.tokens.delete(token)
        this.revokedTokens.add(token)
      }
    }
  }

  /**
   * Révoke tous les tokens expirés
   */
  revokeExpiredTokens() {
    const now = Date.now()
    for (const [token, data] of this.tokens.entries()) {
      if (now > data.expiresAt) {
        this.tokens.delete(token)
        this.revokedTokens.add(token)
      }
    }
  }

  /**
   * Renouvelle un token (rotation)
   */
  rotateToken(oldToken, clientId = null, metadata = {}) {
    if (!this.isValid(oldToken)) {
      throw new Error('Invalid token for rotation')
    }

    const oldTokenData = this.tokens.get(oldToken)
    const newToken = this.issueToken(
      clientId || oldTokenData.clientId,
      { ...oldTokenData.metadata, ...metadata },
      oldTokenData.expiresAt - Date.now()
    )

    // Révoker l'ancien token
    this.revokeToken(oldToken)

    return newToken
  }

  /**
   * Obtient les informations sur un token
   */
  getTokenInfo(token) {
    if (!this.isValid(token)) {
      return null
    }

    const data = this.tokens.get(token)
    return {
      issuedAt: data.issuedAt,
      expiresAt: data.expiresAt,
      clientId: data.clientId,
      metadata: data.metadata,
      lastUsed: data.lastUsed,
      expiresIn: data.expiresAt - Date.now(),
    }
  }

  /**
   * Nettoie les tokens expirés et les anciennes révocations
   */
  cleanup() {
    const now = Date.now()

    // Supprimer les tokens expirés
    for (const [token, data] of this.tokens.entries()) {
      if (now > data.expiresAt) {
        this.tokens.delete(token)
      }
    }

    // Nettoyer les anciennes révocations (garder seulement les 1000 dernières)
    if (this.revokedTokens.size > 1000) {
      const tokensArray = Array.from(this.revokedTokens)
      this.revokedTokens = new Set(tokensArray.slice(-1000))
    }
  }

  /**
   * Génère un token aléatoire sécurisé
   */
  generateToken() {
    return crypto.randomBytes(32).toString('hex')
  }

  /**
   * Obtient les statistiques
   */
  getStats() {
    const now = Date.now()
    let active = 0
    let expired = 0

    for (const data of this.tokens.values()) {
      if (now > data.expiresAt) {
        expired++
      } else {
        active++
      }
    }

    return {
      active,
      expired,
      revoked: this.revokedTokens.size,
      total: this.tokens.size,
    }
  }

  /**
   * Ferme le gestionnaire (arrête le nettoyage automatique)
   */
  close() {
    if (this.cleanupInterval) {
      clearInterval(this.cleanupInterval)
      this.cleanupInterval = null
    }
    this.tokens.clear()
    this.revokedTokens.clear()
  }
}

// Singleton pour le serveur
let globalTokenManager = null

export function getTokenManager(defaultTtl = 24 * 60 * 60 * 1000) {
  if (!globalTokenManager) {
    globalTokenManager = new TokenManager(defaultTtl)
  }
  return globalTokenManager
}

export function createTokenManager(defaultTtl = 24 * 60 * 60 * 1000) {
  return new TokenManager(defaultTtl)
}

export default TokenManager

File: server/backend/app/utils/token_manager.py
from __future__ import annotations

import secrets
import time
from dataclasses import dataclass
from typing import Dict, Optional


@dataclass
class TokenData:
    token: str
    client_id: Optional[str]
    metadata: Dict
    issued_at: float
    expires_at: float


class TokenManager:
    def __init__(self, ttl_seconds: int = 24 * 3600, default_token: Optional[str] = None):
        self.ttl_seconds = ttl_seconds
        self.default_token = default_token
        self._tokens: Dict[str, TokenData] = {}
        if default_token:
            self._tokens[default_token] = TokenData(
                token=default_token,
                client_id=None,
                metadata={"default": True},
                issued_at=time.time(),
                expires_at=float("inf"),
            )

    def issue_token(self, client_id: Optional[str], metadata: Optional[Dict] = None) -> TokenData:
        token = secrets.token_urlsafe(32)
        now = time.time()
        data = TokenData(
            token=token,
            client_id=client_id,
            metadata=metadata or {},
            issued_at=now,
            expires_at=now + self.ttl_seconds,
        )
        self._tokens[token] = data
        return data

    def revoke_token(self, token: str) -> bool:
        if token == self.default_token:
            return False
        return self._tokens.pop(token, None) is not None

    def is_valid(self, token: Optional[str]) -> bool:
        if not token:
            return False
        if token == self.default_token:
            return True
        data = self._tokens.get(token)
        if not data:
            return False
        if data.expires_at < time.time():
            self._tokens.pop(token, None)
            return False
        return True

    def get_info(self, token: str) -> Optional[TokenData]:
        data = self._tokens.get(token)
        if not data:
            return None
        if data.expires_at < time.time():
            self._tokens.pop(token, None)
            return None
        return data

    def rotate_token(self, old_token: str, client_id: Optional[str], metadata: Optional[Dict] = None) -> TokenData:
        if not self.is_valid(old_token):
            raise ValueError("invalid token")
        self.revoke_token(old_token)
        return self.issue_token(client_id, metadata)

    def cleanup(self) -> None:
        now = time.time()
        expired = [t for t, d in self._tokens.items() if d.expires_at < now and t != self.default_token]
        for t in expired:
            self._tokens.pop(t, None)

    def stats(self) -> Dict:
        now = time.time()
        active = [t for t, d in self._tokens.items() if d.expires_at >= now]
        return {
            "active": len(active),
            "default": bool(self.default_token),
        }


# Singleton helper to share the manager across modules
_singleton: Optional[TokenManager] = None


def get_token_manager(default_token: Optional[str] = None, ttl_seconds: int = 24 * 3600) -> TokenManager:
    global _singleton
    if _singleton is None:
        _singleton = TokenManager(ttl_seconds=ttl_seconds, default_token=default_token)
    return _singleton
